src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs

```diff
@@ -0,0 +1,170 @@
+        }
+
+        internal IEnumerable<MissingReference> Compare(IDictionary<string, IEnumerable<string>> nuspecReferences, IEnumerable<string> refFiles)
+        {
+            var missingItems = new List<MissingReference>();
+            if (nuspecReferences.Count() != 0)
+            {
+                if (refFiles.Count() != 0)
+                {
+                    var filesByTFM = refFiles.Where(t => t.Count(m => m == '/') > 1).GroupBy(t => NuGetFramework.ParseFolder(t.Split('/')[1]).GetShortFolderName(), t => Path.GetFileName(t));
+                    foreach (var files in filesByTFM)
+                    {
+                        if (files.Key == "unsupported")
+                        {
+                            continue;
+                        }
+
+                        string[] missingReferences;
+                        string[] missingFiles;
+                        IEnumerable<string> anyReferences = null;
+                        if (nuspecReferences.TryGetValue(files.Key, out var currentReferences) || nuspecReferences.TryGetValue("any", out anyReferences))
+                        {
+                            if (anyReferences != null && currentReferences == null)
+                            {
+                                missingReferences = files.Where(m => !anyReferences.Contains(m)).ToArray();
+                                missingFiles = anyReferences.Where(t => !files.Contains(t)).ToArray();
+                            }
+                            else
+                            {
+                                missingReferences = files.Where(m => !currentReferences.Contains(m)).ToArray();
+                                missingFiles = currentReferences.Where(t => !files.Contains(t)).ToArray();
+                            }
+                        }
+                        else
+                        {
+                            missingReferences = files.ToArray();
+                            missingFiles = Array.Empty<string>();
+                        }
+
+                        if (missingFiles.Length != 0)
+                        {
+                            missingItems.Add(new MissingReference("ref", files.Key, missingFiles));
+                        }
+
+                        if (missingReferences.Length != 0)
+                        {
+                            missingItems.Add(new MissingReference("nuspec", files.Key, missingReferences));
+                        }
+                    }
+                }
+                else
+                {
+                    foreach (var item in nuspecReferences)
+                    {
+                        missingItems.Add(new MissingReference("ref", item.Key, item.Value.ToArray()));
+                    }
+                }
+            }
+            return missingItems;
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(IEnumerable<MissingReference> missingItems)
```

Comment:
the variable name doesn't match the type name, but does match a property name of the type, which is a bit confusing.